src/python/pants/util/objects.py

```diff
@@ -229,41 +297,37 @@ def create(cls, value=None):
 
     def __new__(cls, *args, **kwargs):
       this_object = super(ChoiceDatatype, cls).__new__(cls, *args, **kwargs)
-
-      field_value = getattr(this_object, field_name)
-
-      cls._check_value(field_value)
-
+      cls._check_value(cls._get_value(this_object))
       return this_object
 
-  return ChoiceDatatype
-
-
-class TypedDatatypeClassConstructionError(Exception):
-
-  # TODO: make some wrapper exception class to make this kind of
-  # prefixing easy (maybe using a class field format string?).
-  def __init__(self, type_name, msg, *args, **kwargs):
-    full_msg =  "error: while trying to generate typed datatype {}: {}".format(
-      type_name, msg)
-    super(TypedDatatypeClassConstructionError, self).__init__(
-      full_msg, *args, **kwargs)
-
-
-class TypedDatatypeInstanceConstructionError(TypeError):
-
-  def __init__(self, type_name, msg, *args, **kwargs):
-    full_msg = "error: in constructor of type {}: {}".format(type_name, msg)
-    super(TypedDatatypeInstanceConstructionError, self).__init__(
-      full_msg, *args, **kwargs)
+    def resolve_for_enum_variant(self, mapping):
+      """Return the object in `mapping` with the key corresponding to the enum value.
+
+      `mapping` is a dict mapping enum variant value -> arbitrary object. All variant values must be
+      provided.
+
+      NB: The objects in `mapping` should be made into lambdas if lazy execution is desired, as this
+      will "evaluate" all of the values in `mapping`.
+      """
+      # Equality between a frozenset() and an OrderedSet() is done without respect to ordering,
+      # which is what we want here. We only maintain an OrderedSet() in self.allowed_values so that
+      # we can present error messages with the same arguments used in the constructor.
+      keys = frozenset(mapping.keys())
+      if keys != self.allowed_values:
+        raise self.make_type_error(
+          "pattern matching must have exactly the keys {} (was: {})"
+          .format(list(self.allowed_values), list(keys)))
+      match_for_variant = mapping[self._get_value(self)]
+      return match_for_variant
 
+  return ChoiceDatatype
 
-class TypeCheckError(TypedDatatypeInstanceConstructionError):
 
-  def __init__(self, type_name, msg, *args, **kwargs):
-    formatted_msg = "type check error:\n{}".format(msg)
-    super(TypeCheckError, self).__init__(
-      type_name, formatted_msg, *args, **kwargs)
+# TODO: allow declaring option type automatically as well?
+def register_enum_option(register, enum_cls, *args, **kwargs):
+  """A helper method for declaring a pants option from an `enum()`."""
+  default_value = kwargs.pop('default', enum_cls.default_value)
+  register(*args, choices=enum_cls.allowed_values, default=default_value, **kwargs)
```

Comment:
I like this step, but it makes me wonder how much work it'd be to allow enums to be a type that can be passed directly to `register`. eg, instead of:

```
register_enum_option(
      register, GlobMatchErrorBehavior, '--glob-expansion-failure', default='warn', type=str,
      advanced=True,
      help="Raise an exception if any targets declaring source files "
           "fail to match any glob provided in the 'sources' argument.")
```
something like this:
```
register('--glob-expansion-failure', advanced=True, 
      type=GlobMatchErrorBehavior, default=GlobMatchErrorBehavior.warn,
      help="Raise an exception if any targets declaring source files "
           "fail to match any glob provided in the 'sources' argument.")
```

Reply:
This feels a lot more idiomatic than adding a specific method, and worth doing here to avoid having to fix things up later.

One other thing to note: you used `GlobMatchErrorBehavior.warn` -- adding these attributes to the class itself hasn't been done yet but is a *fantastic* idea. This can be added easily as well.

Reply:
There were some difficulties with implementing this immediately, but I have split those changes out into #7232 and #7233!